chain/client/src/sync/state/mod.rs

```diff
@@ -0,0 +1,388 @@
+mod util;
+
+use crate::metrics;
+use crate::sync::external::{create_bucket_readonly, ExternalConnection};
+use downloader::StateSyncDownloader;
+use external::StateSyncDownloadSourceExternal;
+use futures::future::BoxFuture;
+use near_async::futures::{FutureSpawner, FutureSpawnerExt};
+use near_async::messaging::AsyncSender;
+use near_async::time::{Clock, Duration};
+use near_chain::types::RuntimeAdapter;
+use near_chain::Chain;
+use near_chain_configs::{ExternalStorageConfig, ExternalStorageLocation, SyncConfig};
+use near_client_primitives::types::{ShardSyncStatus, StateSyncStatus};
+use near_epoch_manager::EpochManagerAdapter;
+use near_network::types::{
+    HighestHeightPeerInfo, PeerManagerMessageRequest, PeerManagerMessageResponse,
+};
+use near_primitives::hash::CryptoHash;
+use near_primitives::network::PeerId;
+use near_primitives::state_sync::{ShardStateSyncResponse, ShardStateSyncResponseHeader};
+use near_primitives::types::ShardId;
+use near_store::Store;
+use network::{StateSyncDownloadSourcePeer, StateSyncDownloadSourcePeerSharedState};
+use shard::{run_state_sync_for_shard, StateSyncShardHandle};
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex};
+use task_tracker::{TaskHandle, TaskTracker};
+use tokio::sync::mpsc::{UnboundedReceiver, UnboundedSender};
+use tokio::sync::oneshot;
+use tokio::sync::oneshot::error::TryRecvError;
+use tokio_util::sync::CancellationToken;
+
+/// Module that manages state sync. Internally, it spawns multiple tasks to download state sync
+/// headers and parts in parallel for the requested shards, but externally, all that it exposes
+/// is a single `run` method that should be called periodically, returning that we're either
+/// done or still in progress, while updating the externally visible status.
+pub struct StateSync {
+    store: Store,
+    future_spawner: Arc<dyn FutureSpawner>,
+    epoch_manager: Arc<dyn EpochManagerAdapter>,
+    runtime: Arc<dyn RuntimeAdapter>,
+
+    /// We keep a reference to this so that peer messages received about state sync can be
+    /// given to the StateSyncDownloadSourcePeer.
+    peer_source_state: Arc<Mutex<StateSyncDownloadSourcePeerSharedState>>,
+
+    /// The main downloading logic.
+    downloader: Arc<StateSyncDownloader>,
+
+    /// Internal parallelization limiters as well as status tracker. We need a handle here to
+    /// export statuses of the workers to the debug page.
+    downloading_task_tracker: TaskTracker,
+    computation_task_tracker: TaskTracker,
+
+    /// These are used to submit tasks that must be performed synchronously on the
+    /// Chain. To achieve that, state sync posts a request to one of these queues
+    /// and then in `run` we process them.
+    header_validation_queue: UnboundedReceiver<StateHeaderValidationRequest>,
+    chain_finalization_queue: UnboundedReceiver<ChainFinalizationRequest>,
+    chain_finalization_sender: UnboundedSender<ChainFinalizationRequest>,
```

Comment:
Why do we need to store the sender as part of StateSync?

Reply:
This is removed; replaced with AsyncSender.